server/blob/blob-objectstorage/src/main/java/org/apache/james/blob/objectstorage/StreamCompatibleBlobPutter.java

```diff
@@ -69,14 +71,18 @@ public StreamCompatibleBlobPutter(BlobStore blobStore) {
     public Mono<BlobId> putAndComputeId(ObjectStorageBucketName bucketName, Blob initialBlob, Supplier<BlobId> blobIdSupplier) {
         return putDirectly(bucketName, initialBlob)
             .then(Mono.fromCallable(blobIdSupplier::get))
-            .map(blobId -> updateBlobId(bucketName, initialBlob.getMetadata().getName(), blobId));
+            .flatMap(blobId -> updateBlobId(bucketName, initialBlob.getMetadata().getName(), blobId));
     }
 
-    private BlobId updateBlobId(ObjectStorageBucketName bucketName, String from, BlobId to) {
-        String bucketNameAsString = bucketName.asString();
-        blobStore.copyBlob(bucketNameAsString, from, bucketNameAsString, to.asString(), CopyOptions.NONE);
-        blobStore.removeBlob(bucketNameAsString, from);
-        return to;
+    private Mono<BlobId> updateBlobId(ObjectStorageBucketName bucketName, String from, BlobId to) {
+        return blobExistenceTester.exists(bucketName, to)
+            .map(exists -> {
+                if (!exists) {
```

Comment:
What is the case where it exists?

Reply:
You are writing a big blob already stored in the blobstore.

Given the input (inputStream) you don't know in advance the blobId but only after writing it to a temporary blob that then gets renamed.

blobExistence tester allow you to skip the (expensive) rename operation (behaviour specific to swift).

I would prefer relying on a temporary file for blobId computation....